### PR TITLE
Add cookbook for 2d annulus pyvista visualization

### PR DESCRIPTION
**Pull request by KerrMadeleine**
I have made a short cookbook that describes how to use the pyvista 2D annulus visualization script [aspect/contrib/python/scripts/pyvista_2d_annulus_example.py]. The parameter file used to make the models that I ran the script on is in the cookbook directory. The output figures (angular, radial velocity and temperature deviation) and a document describing how the script works are also available in the doc/ sub-directory.

### CODE REVIEW

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+(sec:cookbooks:2d_annulus_visualization)=
+# Two-dimensional cylindrical annulus visualization
+
+*This section was contributed by Madeleine Kerr.*
+
+This cookbook describes how to use the Python package pyvista to read in time series data from ASPECT .pvd and .pvtu output files(commonly written as 'solution.pvd' or 'solution-00000.pvtu' within the automatically generated [output/solution/] folder). This particular script has the build-in assumption that the model has a 2D cylindrical annulus geometry since it extracts radial and angular velocity from x,y coordinates and velocities and uses the inner and outer radius to threshold the mesh and perform operations.
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:6]: missing space between 'files' and '(commonly...)'

Comment (cedrict) [reply cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:6]: build-in or built-in?

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+The input file to generate the model that created the accompanying figures can be found at[cookbooks/twoD_annulus_visualization/2D_annulus_example.prm]. The ASPECT parameter file is a simple modification of the onset_of_convection cookbook contributed by Juliane Dannberg.
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:8]: no space between 'at' and '[cookbook...]

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+## How to get started
+Since this Python script utilizes pyvista, you should install this package using anaconda following the instructions in the README.md file found in the directory [/aspect/contrib/python/]. Other python scripts that contribute to ASPECT's  functionality can be found within that directory, as well as another 2D pyvista visualization tutorial for a box geometry called 'pyvista_2d_example.py', contributed by John Naliboff.
+
+More information about installing Python via anaconda can be found here: https://docs.anaconda.com/free/anaconda/install/index.html
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:13]: could you make the link a hyperlink?

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+## What the script does
+
+This script can be treated a template and modified based on the fields you are interested in examining. A simple pyvista plot to show and not save the image has the basic form:
+
+    pl = pv.Plotter()
+    pl.add_mesh(
+        mesh,
+        scalars="p"
+        clim=[0,1e11], # color bar range from 0 to 100 GPa
+        cmap="RdBu",   # color map to use
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:46]: may be after this you could add a sentence or two about colormaps. 
see https://www.fabiocrameri.ch/colourmaps/
and this article why it matters:
@article{crsh20,
author={Crameri, Fabio and Shephard, Grace E. and Heron, Philip J.},
title={The misuse of colour in science communication},
journal={Nature Communications},
pages={5444},
volume={11},
doi={10.1038/s41467-020-19160-7},
year={2020}}

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        lighting=False, # lighting is useful mainly to show shadows on 3D meshes
+                        # but we are plotting in 2D
+    )
+    pl.view_xy()
+    pl.show()
+    pl.clear()
+    pl.close()
+
+Adding the mesh titled "mesh" and informing the plotter which scalar value to plot on that mesh(in this case the pressure field which is labeled "p") is the minimum you need to plot using pyvista.
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:55]: missing space after 'mesh'

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+In the script we plot the field "v_phi", the angular component of the velocity field, which is not an output of ASPECT. To create this new point field, the script performs several operations to:
+
+    1. Extract the x and y coordinates of the mesh points. We do this in order to do operations on those values individually since the mesh points are normally a Nxd array where N is the number of mesh points and d is the dimension. We transpose the array (now dxN), and index the first two values as the x and y coordinate arrays respectively. We save the coordinates as new point fields of the mesh called "x" and "y", which are Nx1 arrays that can be operated on and with other Nx1 arrays such temperature, and they can be plotted as a scalar field.
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:59]: if you look at the rendering of this on readthedocs, you will see that it is not very convenient. Could you make sure that the reader need not scroll to the right in order to read this?

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    2. Extract the x and y components of the velocity vectors at every point on the mesh in a similar way as extracting x and y coordinates.
+    3. We can use the new point arrays "x", "y", "v_x" and "v_y", which are Nx1 numpy array objects, to compute the components of the velocity in polar and radial coordinates and plot those fields.
+   
+The radial component of the velocity field from cartesian coordinates is
+```{math}
+v_r = \frac{x\dot{x}+y\dot{y}}{\sqrt{x^2+y^2}},
```
Comment (cedrict) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:65]: may be add afterwards where \dot{x} is the x-component of the velocity. etc ... ?

**Review by cedrict — COMMENTED** [review]
thank you for contributing this. I have left a few comments/remarks.

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md (hunk 4 above)
Comment (KerrMadeleine) [reply cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:46]: Do you recommend I change RdBu to Crameri's vik? I usually use Crameri color maps, however I neglected to add it to my script (perhaps to avoid another python package for a user to install, no real reason)?

Comment (cedrict) [reply cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md]: not necessary, with the new paragraph you added about colormaps.

**Comment by cedrict** [comment]
I have no more comments.

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
+(sec:cookbooks:2d_annulus_visualization)=
+# Two-dimensional cylindrical annulus visualization
+
+*This section was contributed by Madeleine Kerr.*
+
+This cookbook describes how to use the Python package pyvista to read in time series data from ASPECT .pvd and .pvtu output files (commonly written as 'solution.pvd' or 'solution-00000.pvtu' within the automatically generated [output/solution/] folder). This particular script has the built-in assumption that the model has a 2D cylindrical annulus geometry since it extracts radial and angular velocity from x,y coordinates and velocities and uses the inner and outer radius to threshold the mesh and perform operations.
+
+The input file to generate the model that created the accompanying figures can be found at [cookbooks/twoD_annulus_visualization/2D_annulus_example.prm]. The ASPECT parameter file is a simple modification of the onset_of_convection cookbook contributed by Juliane Dannberg.
```
Comment (gassmoeller) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:8]: It is very nice that you attribute the individual files to their authors, however we handle authorship citations within the ASPECT project slightly differently. You receive credit for your contribution by adding a changelog entry signed with your name when you contribute a larger piece of code. Potentially you will also be added to the list of authors of the manual, and/or the principal developers of the software. However, all code/prms/manual entries that are contributed to ASPECT become part of the software project as a whole, and are usually not attributed to a single person (except for the line at the top of this file). We chose this convention to make clear that every piece of ASPECT is part of the community code, and everyone has the right to change or improve every piece of ASPECT.

So TLDR: Please do not list the author of every other file in ASPECT that you mention. But please add a changelog entry under your name to `doc/modules/changes`.

```suggestion
The input file to generate the model that created the accompanying figures can be found at [cookbooks/2D_annulus_visualization/2D_annulus_example.prm]. The ASPECT parameter file is a simple modification of the onset_of_convection cookbook.
```

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+## How to get started
+Since this Python script utilizes pyvista, you should install this package using anaconda following the instructions in the README.md file found in the directory [/aspect/contrib/python/]. Other python scripts that contribute to ASPECT's  functionality can be found within that directory, as well as another 2D pyvista visualization tutorial for a box geometry called 'pyvista_2d_example.py', contributed by John Naliboff.
```
Comment (gassmoeller) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:11]:
```suggestion
Since this Python script utilizes pyvista, you should install this package using anaconda following the instructions in the README.md file found in the directory [/aspect/contrib/python/]. Other python scripts that contribute to ASPECT's  functionality can be found within that directory, as well as another 2D pyvista visualization tutorial for a box geometry called 'pyvista_2d_example.py'.
```

File: cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    out_n_min=0 # The initial solution file to visualize: i.e. 'solution-00000.pvtu'
+    out_n_max=25 # The final solution file to visualize: i.e. 'solution-00025.pvtu'
+
+and the inner and outer radius of the annulus from the geometry of your model:
+    r_inner = 3481000 # units in m
+    r_outer = 6371000 # units in m.
+
+## What the script does
+
+This script can be treated a template and modified based on the fields you are interested in examining. A simple pyvista plot to show and not save the image has the basic form:
```
Comment (gassmoeller) [inline cookbooks/2d_annulus_visualization/doc/2d_annulus_visualization.md:39]:
```suggestion
This script can be treated as a template and modified based on the fields you are interested in examining. A simple pyvista plot to show and not save the image has the basic form:
```

File: cookbooks/2d_annulus_visualization/2d_annulus_example.prm [hunk 11]
```diff
@@ -0,0 +1,89 @@
+# This setup is a copy of the
+# cookbook/onset-of-convection/onset-of-convection.prm
+# contributed by Max Rudolph and Juliane Dannberg, with the difference that
```
Comment (gassmoeller) [inline cookbooks/2d_annulus_visualization/2d_annulus_example.prm:3]: See my comment below.

```suggestion
# with the difference that
```

**Review by gassmoeller — COMMENTED** [review]
Very close. Only remaining comments:

- A few minor comments.
- Please add a changelog entry to `doc/modules/changes`
- Please add a test that is a simpler and faster version of this cookbook (talk to me about how to avoid duplicating the .prm file for the test). This test makes sure the cookbook will continue to run in the future.

**Review by gassmoeller — APPROVED** [review]
